### PR TITLE
Fix joining with an inherited property from another table for Linq provider

### LINKED ISSUE #2580: InvalidWithClauseException when join polymorphic entity
### Situation
Updating NHibernate from 5.2.7 to 5.3.3 we ran into issues with queries operating on inheritance hierarchies. We join inherited entities and run operators such as distincts or simple selects. Analyzing the exception `with clause can only reference columns in the driving table` we soon came to find the cause being the `HqlSqlWalker` and its `WithClauseVisitor`. There the names of aliases are compared and if different, an exception is thrown.

#### Comparison with Hibernate
After some research we found this to be an issue in Hibernate as well: https://hibernate.atlassian.net/browse/HHH-11383. It was fixed here: https://github.com/hibernate/hibernate-orm/commit/5ac46eb70359a45a34f7a64e859df106928baa94. 

We use NHibernate as our ORM for large-scale projects and ensure compability with a test coverage of 20K+ tests (hence we discovered this issue quickly). For testing purposes we disabled the exception (as in Hibernate) and ran our queries, which then work smoothly. 

### Request
Is it possible to adapt the change from Hibernate to NHibernate and disable the check of alias names in this case and consequently not throwing the exception?

On a side note: perhaps this became a problem in the first place after NHibernate release 5.3.0 and this change https://github.com/nhibernate/nhibernate-core/pull/2078

### PR DESCRIPTION
**Pull request by maca88**
Fix for a regression made by #2327. When merging into master only the test cases should be merged as #2361 makes this fix obsolete.

Fixes #2580

### CODE REVIEW

File: src/NHibernate/Linq/Visitors/QueryModelVisitor.cs [hunk 1]
```diff
@@ -527,10 +529,13 @@ private void AddJoin(JoinClause joinClause, QueryModel queryModel, bool innerJoi
 			var withClause = equalityVisitor.Visit(joinClause.InnerKeySelector, joinClause.OuterKeySelector);
 			var alias = _hqlTree.TreeBuilder.Alias(VisitorParameters.QuerySourceNamer.GetName(joinClause));
 			var joinExpression = HqlGeneratorExpressionVisitor.Visit(joinClause.InnerSequence, VisitorParameters);
+			var baseMemberCheker = new BaseMemberChecker(VisitorParameters.SessionFactory);
+
 			HqlTreeNode join;
-			// When associations are located inside the inner key selector we have to use a cross join instead of an inner
```
Comment (bahusoid) [inline src/NHibernate/Linq/Visitors/QueryModelVisitor.cs:531]: A bit offtopic: Is this check still required in master? Or it also should be handled by #2361 and can be removed?

File: src/NHibernate/Linq/Visitors/QueryModelVisitor.cs [hunk 2]
```diff
@@ -551,6 +556,46 @@ private void AddJoin(JoinClause joinClause, QueryModel queryModel, bool innerJoi
...
+			protected override Expression VisitMember(MemberExpression node)
+			{
+				if (ExpressionsHelper.TryGetMappedType(
+						_sessionFactory,
+						node,
+						out _,
+						out var persister,
+						out _,
+						out var propertyPath) &&
+					persister is IOuterJoinLoadable joinLoadable &&
```
Comment (bahusoid) [inline src/NHibernate/Linq/Visitors/QueryModelVisitor.cs:586]: Maybe skip property checks if no subclasses? Something like
```C#
persister.EntityMetamodel.HasSubclasses &&
persister is IOuterJoinLoadable joinLoadable ...
```

Or skip this visitor completely if `joinClause` entity has no subclasses?

File: src/NHibernate/Linq/Visitors/QueryModelVisitor.cs (hunk 1 above)
Comment (maca88) [reply src/NHibernate/Linq/Visitors/QueryModelVisitor.cs:531]: No, it was obsolete by #2539. We should remove the `ParentJoinClause` property on the master branch as now is never set.

File: src/NHibernate/Linq/Visitors/QueryModelVisitor.cs (hunk 2 above)
Comment (maca88) [reply src/NHibernate/Linq/Visitors/QueryModelVisitor.cs:594]: > Or skip this visitor completely if joinClause entity has no subclasses?

Added a check before visiting the node.

File: src/NHibernate/Linq/Visitors/QueryModelVisitor.cs [hunk 3]
```diff
@@ -551,6 +556,54 @@ private void AddJoin(JoinClause joinClause, QueryModel queryModel, bool innerJoi
...
+
+			public BaseMemberChecker(ISessionFactoryImplementor sessionFactory)
+			{
+				_sessionFactory = sessionFactory;
+			}
+
+			public bool ContainsBaseMember(JoinClause joinClause)
+			{
+				// Visit the join inner key only for entities that have subclasses
+				if (joinClause.InnerSequence is ConstantExpression constantNode &&
```
Comment (maca88) [inline src/NHibernate/Linq/Visitors/QueryModelVisitor.cs:572]: Constant check seems to be enough as `join o2 in db.Mammals.Where(o => o.Pregnant)` is not supported.

**Review by bahusoid — APPROVED** [review]

**Pull request merged**